### PR TITLE
♻️🔒 Use unique ids for UserTask events

### LINKED ISSUE #69: ⚡️Finishing UserTasks creates a large number of database requests
#### Describe your issue

See Issue: https://github.com/process-engine/process_engine_runtime/issues/124

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

## Example

<details>
<summary>
:warning: Before creating the issue, click here to expand and follow the guide.
</summary>
<br>
  
1. Be sure to fill out **every** topic presented in this template. 
If a section is not valid or useful for your issue, feel free to remove it.

2. Use the attached Checklist to make sure your issue contains all relevant information.

3. Attach a label to your issue that reflects what your issue is all about.
**When your issue concerns only a specific platform, be sure to attach the label for that platform!**

4. Please use a fitting emoji at the beginning of your issue title.

Example 1 - Report a bug:
```md
:bug: Something is not working properly
```

Example 2 - Report a performance issue:
```md
:zap: Something is slower than it should be
```

Example 3 - Report a package vulnerability or security issue:
```md
:lock: Package xy produces vulnerabilites
```

```md
:lock: User XY has access to something he shouldn't have
```

Example 4 - Report missing test coverage: 
```md
:white_check_mark: Add tests for...
```

You can of course phrase your titles however you like.
Just be sure they are short, to the point and convey the general idea of the issue you are experiencing.
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your Issue with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### LINKED ISSUE #70: ♻️ Use instance IDs for sending/receiving UserTask messages
#### Describe your issue

Relates to: https://github.com/process-engine/process_engine_core/issues/187

We currently use the following IDs when creating Events and Event-Subscriptions for UserTasks:
- CorrelationId
- ProcessmodelId
- UserTaskId

The problem with this approach is that none of these values are necessarily unique. 
This means that we could run into an issue, where a single command for finishing a UserTask can actually finish more than one.

#### Possible solution

We need at least one value in this chain that is guaranteed to be unique.
This can be the ProcessModel's `ProcessInstanceId` and/or the UserTask's `FlowNodeInstanceId`.

Then we'd have events like this:
`correlation/:someCorrelationId/processinstance/:someProcessInstanceId/usertaskinstance/:someFlowNodeInstanceId/finish`.

This would be the most secure solution, as every value - except for the CorrelationId - is guaranteed to be unique.

It would also be possible to just use either the ProcessInstanceId or the UserTaskId, but I think using both provides greater benefit in terms of data consistency and security.

**NOTE:** 
Of course, this means that the User performing the reqeust against the Consumer API (BPMN Studio, etc) need to provide those unique IDs with their request aswell. Otherwise, the Consumer API would need to query the database, using the UserTaskId and ProcessModelId and we would be right back where we started.

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

## Example

<details>
<summary>
:warning: Before creating the issue, click here to expand and follow the guide.
</summary>
<br>
  
1. Be sure to fill out **every** topic presented in this template. 
If a section is not valid or useful for your issue, feel free to remove it.

2. Use the attached Checklist to make sure your issue contains all relevant information.

3. Attach a label to your issue that reflects what your issue is all about.
**When your issue concerns only a specific platform, be sure to attach the label for that platform!**

4. Please use a fitting emoji at the beginning of your issue title.

Example 1 - Report a bug:
```md
:bug: Something is not working properly
```

Example 2 - Report a performance issue:
```md
:zap: Something is slower than it should be
```

Example 3 - Report a package vulnerability or security issue:
```md
:lock: Package xy produces vulnerabilites
```

```md
:lock: User XY has access to something he shouldn't have
```

Example 4 - Report missing test coverage: 
```md
:white_check_mark: Add tests for...
```

You can of course phrase your titles however you like.
Just be sure they are short, to the point and convey the general idea of the issue you are experiencing.
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your Issue with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### PR DESCRIPTION
**Pull request by ElRaptorus**
**Changes:**

1. Add the `FlowNodeInstanceId` to the UserTask datatype.
2. Change the parameter names for the `finishUserTask` UseCase to indicate InstanceIds, not ModelIds.
3. Refactor `finishUserTask` UseCase:
    * Split the UseCase into smaller tasks and move each into its own function.
    * Avoid using the `getUserTasksForProcessModelInCorrelation` to avoid side effects, like getting more UserTasks than you wanted to.
4. Remove the `_checkIfCorrelationExists` and `_checkIfProcessInstanceExists` checks, since they each performed additional database queries that seriously impacted performance.
    * When a `getUserTask` request is made and the given CorrelationId and/or ProcessInstance does not exists, an empty Array is returned now, instead of a 404.
5. Finish UserTasks by the given InstanceIds, instead of the given ModelIds, to ensure that only one UserTask will ever be finished per request.

**Issues:**

Part of https://github.com/process-engine/process_engine_runtime/issues/94
Part of https://github.com/process-engine/process_engine_runtime/issues/124
Part of https://github.com/process-engine/process_engine_runtime/issues/126

Closes #69 
Closes #70 

PR: #71

## How can others test the changes?

Use the Consumer API to handle and finish UserTask, using InstanceIds, instead of ModelIds.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Fixing Security Issues   | `:lock:`             | 🔒     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️      |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️      |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/consumer_api_service.ts [hunk 1]
```diff
@@ -363,7 +354,36 @@ export class ConsumerApiService implements IConsumerApi {
...
+    const matchingUserTask: UserTask =
+      userTaskList.userTasks.find((userTask: UserTask): boolean => {
+        return userTask.flowNodeInstanceId === userTaskInstanceId
+          && userTask.processInstanceId === processInstanceId;
+      });
+
+    const noMatchingUserTaskFound: boolean = matchingUserTask === undefined;
+    if (noMatchingUserTaskFound) {
+      const errorMessage: string =
+        `ProcessInstance '${processInstanceId}' in Correlation '${correlationId}' does not have a UserTask '${userTaskInstanceId}'`;
```
Comment (robpal94) [inline src/consumer_api_service.ts]: I would expect something like _...does not have a UserTask with an Id..._ or simply no output of an id.

File: src/converters/user_task_converter.ts [hunk 2]
```diff
@@ -59,34 +60,21 @@ export class UserTaskConverter {
...
+      const flowNodeModel: Model.Base.FlowNode = processModelFacade.getFlowNodeById(suspendedFlowNode.flowNodeId);
 
-      const userTask: UserTask =
-        await this.convertSuspendedFlowNodeToUserTask(suspendedFlowNode, currentProcessToken, processModel);
+      // Note that UserTasks are not the only types of FlowNodes that can be suspended.
+      // So we must make sure that what we have here is actually a UserTask and not, for example, a TimerEvent.
+      const flowNodeIsAUserTask: boolean = flowNodeModel.constructor.name === 'UserTask';
 
-      if (userTask === undefined) {
+      if (!flowNodeIsAUserTask) {
```
Comment (Vyperus) [inline src/converters/user_task_converter.ts]: Don't use negations in the if-clause to increase readability.

File: src/converters/user_task_converter.ts [hunk 3]
```diff
@@ -59,34 +60,21 @@ export class UserTaskConverter {
...
-        processModel = await this.processModelService.getProcessModelById(identity, currentProcessToken.processModelId);
-        processModelCache[currentProcessToken.processModelId] = processModel;
-      }
+      const flowNodeModel: Model.Base.FlowNode = processModelFacade.getFlowNodeById(suspendedFlowNode.flowNodeId);
 
-      const userTask: UserTask =
-        await this.convertSuspendedFlowNodeToUserTask(suspendedFlowNode, currentProcessToken, processModel);
+      // Note that UserTasks are not the only types of FlowNodes that can be suspended.
+      // So we must make sure that what we have here is actually a UserTask and not, for example, a TimerEvent.
+      const flowNodeIsAUserTask: boolean = flowNodeModel.constructor.name === 'UserTask';
```
Comment (Vyperus) [inline src/converters/user_task_converter.ts]:
```suggestion
      const flowNodeIsNotAUserTask: boolean = flowNodeModel.constructor.name !== 'UserTask';
```

File: src/converters/user_task_converter.ts (hunk 2 above)
Comment (Vyperus) [inline src/converters/user_task_converter.ts]:
```suggestion
      if (!flowNodeIsNotAUserTask) {
```

**Comment by ElRaptorus** [comment]
@Vyperus done

**Pull request merged**